qiita_db/portal.py

```diff
@@ -5,13 +5,66 @@
 #
 # The full license is in the file LICENSE, distributed with this software.
 # -----------------------------------------------------------------------------
+import warnings
 
 from .sql_connection import SQLConnectionHandler
 from .util import convert_to_id
+from .exceptions import QiitaDBError, QiitaDBDuplicateError
 from qiita_core.qiita_settings import qiita_config
+from qiita_core.exceptions import IncompetentQiitaDeveloperError
 
 # Only make functions available for main portal
 if qiita_config.portal == "QIITA":
+    def create_portal(portal, desc):
+        """Creates a new portal on the system
+
+        Parameters
+        ----------
+        portal : str
+            The name of the portal to add
+        desc : str
+            Description of the portal
+
+        Raises
+        ------
+        QiitaDBDuplicateError
+            Portal already exists
+        """
+        try:
+            convert_to_id(portal, 'portal_type', 'portal')
+        except IncompetentQiitaDeveloperError:
+            # Portal does not exist, so add it
+            sql = """INSERT INTO qiita.portal_type (portal, portal_description)
+                     VALUES (%s, %s)"""
+            conn_handler = SQLConnectionHandler()
+            conn_handler.execute(sql, [portal, desc])
+        else:
+            # Portal already exists, so raise error
+            raise QiitaDBDuplicateError("Portal", portal)
+
+    def remove_portal(portal):
+        """Removes a portal from the system
+
+        Parameters
+        ----------
+        portal : str
+            The name of the portal to add
+        desc : str
+            Description of the portal
+
+        Raises
+        ------
+        QiitaDBError
+            Portal has analyses or studies attached to it
+        """
+        sql = "DELETE FROM qiita.portal_type WHERE portal = %s"
+        conn_handler = SQLConnectionHandler()
+        try:
+            conn_handler.execute(sql, [portal])
```

Comment:
Ah, so it is, but we should really rethink that. You can get some nicely fine-grained stuff for errors in psycopg2 Details [here](http://initd.org/psycopg/docs/module.html#psycopg2.InterfaceError) and [here](http://initd.org/psycopg/docs/errorcodes.html). This will also allow us to leverage the errors for foreign key checks, example [here](http://stackoverflow.com/a/12495497).